backends/xsmm/ceed-xsmm-tensor.c

```diff
@@ -0,0 +1,79 @@
+// Copyright (c) 2017-2022, Lawrence Livermore National Security, LLC and other CEED contributors.
+// All Rights Reserved. See the top-level LICENSE and NOTICE files for details.
+//
+// SPDX-License-Identifier: BSD-2-Clause
+//
+// This file is part of CEED:  http://github.com/ceed
+
+#include <ceed.h>
+#include <ceed/backend.h>
+#include <libxsmm.h>
+
+#include "ceed-xsmm.h"
+
+//------------------------------------------------------------------------------
+// Tensor Contract Apply
+//------------------------------------------------------------------------------
+static int CeedTensorContractApply_Xsmm(CeedTensorContract contract, CeedInt A, CeedInt B, CeedInt C, CeedInt J, const CeedScalar *restrict t,
+                                        CeedTransposeMode t_mode, const CeedInt add, const CeedScalar *restrict u, CeedScalar *restrict v) {
+  Ceed ceed;
+  CeedCallBackend(CeedTensorContractGetCeed(contract, &ceed));
+
+  if (C == 1) {
+    // Build or query the required kernel
+    const int                  flags_t    = LIBXSMM_GEMM_FLAGS(!t_mode ? 'T' : 'N', 'N');
+    const int                  flags_ab   = (!add) ? LIBXSMM_GEMM_FLAG_BETA_0 : 0;
+    const int                  flags      = (flags_t | flags_ab);
+    const libxsmm_gemm_shape   gemm_shape = (CEED_SCALAR_TYPE != CEED_SCALAR_FP32)
+                                                ? libxsmm_create_gemm_shape(J, A, B, !t_mode ? B : J, B, J, LIBXSMM_DATATYPE_F64, LIBXSMM_DATATYPE_F64,
+                                                                            LIBXSMM_DATATYPE_F64, LIBXSMM_DATATYPE_F64)
+                                                : libxsmm_create_gemm_shape(J, A, B, !t_mode ? B : J, B, J, LIBXSMM_DATATYPE_F32, LIBXSMM_DATATYPE_F32,
+                                                                            LIBXSMM_DATATYPE_F32, LIBXSMM_DATATYPE_F32);
+    const libxsmm_gemmfunction kernel = libxsmm_dispatch_gemm_v2(gemm_shape, (libxsmm_bitfield)(flags), (libxsmm_bitfield)LIBXSMM_GEMM_PREFETCH_NONE);
+    CeedCheck(kernel, ceed, CEED_ERROR_BACKEND, "LIBXSMM kernel failed to build.");
+
+    // Run kernel
+    libxsmm_gemm_param gemm_param;
+    gemm_param.a.primary = (CeedScalar *)&t[0];
+    gemm_param.b.primary = (CeedScalar *)&u[0];
+    gemm_param.c.primary = (CeedScalar *)&v[0];
+    kernel(&gemm_param);
+  } else {
+    // Build or query the required kernel
+    const int                  flags_t    = LIBXSMM_GEMM_FLAGS('N', t_mode ? 'T' : 'N');
+    const int                  flags_ab   = (!add) ? LIBXSMM_GEMM_FLAG_BETA_0 : 0;
+    const int                  flags      = (flags_t | flags_ab);
+    const libxsmm_gemm_shape   gemm_shape = (CEED_SCALAR_TYPE != CEED_SCALAR_FP32)
```

Comment:
minor - I generally find "if this then that else other" easier to read than "if not this then other else that"
```suggestion
    const libxsmm_gemm_shape   gemm_shape = (CEED_SCALAR_TYPE == CEED_SCALAR_FP64)
```

Reply:
Done: 684fdff.